File: Wire-iOS/Sources/UserInterface/Components/MediaAsset.m
// 
// Wire
// Copyright (C) 2016 Wire Swiss GmbH
// 
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
// 
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
// 
// You should have received a copy of the GNU General Public License
// along with this program. If not, see http://www.gnu.org/licenses/.
// 


#import "MediaAsset.h"


#import <ImageIO/ImageIO.h>
#import <MobileCoreServices/MobileCoreServices.h>
#import "Wire-Swift.h"

@implementation FLAnimatedImage(MediaAsset)

- (BOOL)isGIF
{
    return YES;
}

- (BOOL)isTransparent
{
    return NO;
}

@end



@implementation UIImageView(MediaAssetView)

- (NSData *)imageData
{
    return self.image.data;
}

- (void)setImageData:(NSData *)imageData
{
    self.image = [[UIImage alloc] initWithData:imageData];
}

+ (instancetype)imageViewWithMediaAsset:(id<MediaAsset>)image;
{
    if ([image isGIF]) {
        FLAnimatedImageView *animatedImageView = [[FLAnimatedImageView alloc] init];
        animatedImageView.animatedImage = image;
        return animatedImageView;
    }
    else {
        return [[UIImageView alloc] initWithImage:[(UIImage *)image downsizedImage]];
    }
}

- (id<MediaAsset>)mediaAsset
{
    return self.image;
}

- (void)setMediaAsset:(id<MediaAsset>)image
{
    if (image == nil) {
        self.image = nil;
    }
    else if (![image isGIF]) {
        self.image = [(UIImage *)image downsizedImage];
    }
}

@end



@implementation FLAnimatedImageView(MediaAssetView)

- (id<MediaAsset>)mediaAsset
{
    return self.animatedImage ?: self.image;
}

- (void)setMediaAsset:(id<MediaAsset>)image
{
    if (image == nil) {
        self.image = nil;
        self.animatedImage = nil;
    }
    else {
        if ([image isGIF]) {
            self.animatedImage = image;
        }
        else {
            self.image = [(UIImage *)image downsizedImage];
        }
    }
}

@end

File: Wire-iOS/Sources/UserInterface/Components/MediaAsset.swift
// 
// Wire
// Copyright (C) 2020 Wire Swiss GmbH
// 
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
// 
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
// 
// You should have received a copy of the GNU General Public License
// along with this program. If not, see http://www.gnu.org/licenses/.
// 

import FLAnimatedImage

protocol MediaAsset: AnyObject {
    var imageData: Data? { get }
    var size: CGSize { get }
    var isGIF: Bool { get }
    var isTransparent: Bool { get }
}


extension MediaAsset {
    var imageView: MediaAssetView {
        if isGIF {
            let animatedImageView = FLAnimatedImageView()
            animatedImageView.animatedImage = self as? FLAnimatedImage

            return animatedImageView
        } else {
            return UIImageView(image: (self as? UIImage)?.downsized())
        }
    }

}


protocol MediaAssetView: UIView {
    var mediaAsset: MediaAsset? { get set }
}

extension MediaAssetView where Self: UIImageView {
    var mediaAsset: MediaAsset? {
        get {
            return image
        }
        set {
            if newValue == nil {
                image = nil
            } else if newValue?.isGIF == false {
                image = (newValue as? UIImage)?.downsized()
            }
        }
    }
}

extension MediaAssetView where Self: FLAnimatedImageView {
    var mediaAsset: MediaAsset? {
        get {
            return animatedImage ?? image
        }

        set {
            if let newValue = newValue {
                if newValue.isGIF == true {
                    animatedImage = newValue as? FLAnimatedImage
                } else {
                    image = (newValue as? UIImage)?.downsized()
                }
            } else {
                image = nil
                animatedImage = nil
            }
        }
    }
}

extension FLAnimatedImage: MediaAsset {
    var imageData: Data? {
        return data
    }

    var isGIF: Bool {
        return true
    }

    var isTransparent: Bool {
        return false
    }
}


extension UIImageView: MediaAssetView {

    var imageData: Data? {
        get {
            return image?.imageData
        }

        set {
            if let imageData = newValue {
                image = UIImage(data: imageData)
            }
        }
    }
}



